scala.wake

```diff
@@ -19,25 +28,47 @@ def myMakeStatePath x =
   | runJobWith       virtualRunner
   | getJobOutput
 
-global def readIvyDepsJSON dir =
-  source "{dir}/ivydependencies.json" | parseJSONFile
-
-# Should be installed by wit
+# Maintain compatibility with Wit 0.12.0 and below
 def coursierBin =
   match (files bloopInstall `.*coursier.*`)
     bin, Nil = myMakeStatePath bin
-    Nil      = makeBadPath (makeError "Coursier not found! Did you run `wit fetch-scala`?")
+    Nil      = fetchCoursier Unit
     l        = makeBadPath (makeError "Multiple coursier executables found! {catWith ", " l}")
 
-# Virtual job that owns all fetched ivy dependencies
-def ivyCacheDeps Unit =
+# Only get files 1 level under workspace root (emulates Wit <=0.12.0 behavior)
+def ivyDepsFiles = sources "." `[^/]*/ivydependencies\.json`
+
+# Virtual job that owns all wit-fetched ivy deps
+target existingIvyCacheDeps Unit =
   def fs = files ivyCache `.*`
   makePlan ("<ivycache>", ivyCache, Nil) Nil
   | setPlanKeep      False
   | setPlanFnOutputs (\_ fs)
   | runJobWith       virtualRunner
   | getJobOutputs
 
+# Job that fetches ivy dependencies
+target ivyCacheDeps Unit =
+  def existing = existingIvyCacheDeps Unit
+  def bin = source "{here}/fetch_ivy_dependencies"
+  def dir = mkdir bloopInstall
+  def cache = mkdir ivyCache
+  def otherDeps = source "{here}/lib/scala-bridge-fetcher_2.12-0.1.0.jar", sources here `.*\.py`
+  def cmd = bin.getPathName, "--scala", dir.getPathName, "--cache", cache.getPathName, map getPathName ivyDepsFiles
+  def allDeps = bin, dir, cache, coursierBin, otherDeps ++ ivyDepsFiles ++ existing
+  def ivycacheOnly l = l | filter (matches `ivycache/.*`)
+  def newIvyCacheDeps = match ivyDepsFiles
+    Nil = Nil
+    _   =
+      makePlan cmd allDeps
+      | editPlanEnvironment ("PYTHONDONTWRITEBYTECODE=1", _)
+      # We need to ensure that we only recognize outputs in ivycache, some bloop artifacts are non-deterministic
+      | editPlanFnOutputs (ivycacheOnly ∘ _)
```

Comment:
Yeah, I guess the issue is that `PlanFnOutputs` is already a function, `List String => List String` so `editPlanFnOutputs` wants a higher-order function that accepts the old one as an argument `(List String => List String) => List String => List String)`.

I choose `∘` intentionally because it is the function composition operator to try to make it clear that that's what's happening here. Do you have a suggestion on a better way to express this? If I rename `ivycacheOnly` to `filesInIvycache` and get rid of the compose operator, I can be more explicit with something like:
```
| editPlanFnOutputs (\fn \arg arg | fn | filesInIvycache)
```
But I think that's strictly less clear 🤷‍♂

Reply:
Could also do
```
| editPlanFnOutputs (_ _ | filesInIvycache)
```
Which I find LOL

Reply:
Or
```
def compose f g x = f (g x)
...
| editPlanFnOutputs (compose filesInIvycache)
```

Reply:
Hmm, this may not be elegant, and especially not a good example of functional programming, but what about folding in the "compose" functionality into `filesInIvycache`? For example, defining that function as

```wake
def filterIvycacheFiles origFn files =
   files | origFn | filter (matches `ivycache/.*`)

...

plan
| editPlanFnOutputs filterIvycacheFiles
```

Hmm, that's really no different than your anonymous function version. I feel like there is just something a little bit unergonomic about it. Maybe we just need a custom editor function that only assumes that you will either prepend or append your filter function. E.g.

```wake
def appendPlanFnOutputs origFn newFn plan =
  plan
  | editPlanFnOutputs (origFn (yourUnicodeCharForComposeThatIRefuseToTypeOut) newFn)

...

plan
| appendPlanFnOutputs filesInIvycache
```

I was also about to suggest a prepend version, but maybe that's not even necessary, since I think the only sensible thing to do here is to chain filters that are commutative, so it shouldn't matter where in the chain you add them.

Reply:
BTW, this is a non-blocking comment, so we don't need to rathole on this.

Reply:
I think your `filterIvycacheFiles` suggestion is the most clear so I'll do that.